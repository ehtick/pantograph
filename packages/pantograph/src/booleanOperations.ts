import {
  cutFiguresLists,
  fuseFiguresLists,
  intersectFiguresLists,
} from "./algorithms/boolean/figureBooleans";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { listOfFigures } from "./utils/listOfFigures";

export function fuse(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    fuseFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function fuseAll(shapes: (Diagram | Figure | Loop)[]): Diagram {
  return shapes.reduce(
    (acc: Diagram, shape: Diagram | Figure | Loop) => fuse(acc, shape),
    new Diagram()
  );
}

export function cut(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    cutFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function intersect(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    intersectFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}
