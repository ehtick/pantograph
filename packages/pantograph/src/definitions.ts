export type Vector = [number, number];

export function isVector(vector: unknown): vector is Vector {
  return Array.isArray(vector) && vector.length === 2;
}
