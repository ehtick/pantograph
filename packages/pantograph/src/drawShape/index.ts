export { drawCircle } from "./drawCircle.js";
export { drawRect } from "./drawRect.js";
export { drawEllipse } from "./drawEllipse.js";
export { drawSVGPath } from "./drawSVGPath.js";
