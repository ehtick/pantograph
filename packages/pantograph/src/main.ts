import type { Vector } from "./definitions";
export { Vector };

export const DEG2RAD = Math.PI / 180;
export const RAD2DEG = 180 / Math.PI;

export function polarToCartesian(r: number, theta: number): Vector {
  const x = Math.cos(theta * DEG2RAD) * r;
  const y = Math.sin(theta * DEG2RAD) * r;
  return [x, y];
}

export function cartesianToPolar([x, y]: Vector): [number, number] {
  const r = Math.sqrt(x * x + y * y);
  const theta = Math.atan2(y, x) * RAD2DEG;
  return [r, theta];
}

export type {
  Diagram,
  Figure,
  Loop,
  Strand,
  Stroke,
  TransformationMatrix,
  BoundingBox,
  Segment,
  Line,
  Arc,
} from "./models/exports";

export { draw } from "./draw";
export {
  // Surface booleans
  fuseAll,
  fuse,
  cut,
  intersect,
  // Strand booleans
  eraseStrand,
  confineStrand,
  // Offset
  offset,
} from "./operations";

export { exportSVG, svgBody } from "./export/svg/exportSVG";

export { exportJSON } from "./export/json/exportJSON";
export { importJSON } from "./import/json/importJSON";
