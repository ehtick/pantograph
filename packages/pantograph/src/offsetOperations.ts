import {
  offsetFigures,
  outlineStrokeFigures,
} from "./algorithms/offsets/offsetFigure";
import { outlineStrand } from "./algorithms/offsets/offsetStroke.js";
import { Strand } from "./models/Strand.js";
import { Diagram } from "./models/Diagram.js";
import { Figure } from "./models/Figure.js";
import { Loop } from "./models/Loop.js";
import { listOfFigures } from "./utils/listOfFigures.js";

export function offset(
  shape: Diagram | Figure | Loop,
  offsetDistance: number
): Diagram {
  return offsetFigures(listOfFigures(shape), offsetDistance);
}

export function outlineStroke(
  shape: Diagram | Figure | Loop | Strand,
  outlineDistance: number,
  { endCap = "round" }: { endCap?: "butt" | "round" } = {}
): Diagram {
  if (shape instanceof Strand) {
    return outlineStrand(shape, outlineDistance, endCap);
  }
  return outlineStrokeFigures(listOfFigures(shape), outlineDistance);
}
