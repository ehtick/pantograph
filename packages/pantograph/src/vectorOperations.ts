import type { Vector } from "./definitions";

export const DEG2RAD = Math.PI / 180;

export const reprVector = (vector: Vector): string => {
  return `[${vector[0]}, ${vector[1]}]`;
};

export const sameVector = (
  [x0, y0]: Vector,
  [x1, y1]: Vector,
  precision = 1e-9
): boolean => {
  return Math.abs(x0 - x1) <= precision && Math.abs(y0 - y1) <= precision;
};

export const add = ([x0, y0]: Vector, [x1, y1]: Vector): Vector => {
  return [x0 + x1, y0 + y1];
};

export const subtract = ([x0, y0]: Vector, [x1, y1]: Vector): Vector => {
  return [x0 - x1, y0 - y1];
};

export const squareLength = ([x, y]: Vector): number => {
  return x * x + y * y;
};

export const length = ([x, y]: Vector): number => {
  return Math.sqrt(squareLength([x, y]));
};

export const scalarMultiply = ([x0, y0]: Vector, scalar: number): Vector => {
  return [x0 * scalar, y0 * scalar];
};

export const squareDistance = (
  [x0, y0]: Vector,
  [x1, y1]: Vector = [0, 0]
): number => {
  return (x0 - x1) ** 2 + (y0 - y1) ** 2;
};

export const distance = (p0: Vector, p1: Vector = [0, 0]): number => {
  return Math.sqrt(squareDistance(p0, p1));
};

export function crossProduct([x0, y0]: Vector, [x1, y1]: Vector): number {
  return x0 * y1 - y0 * x1;
}

export function dotProduct([x0, y0]: Vector, [x1, y1]: Vector): number {
  return x0 * x1 + y0 * y1;
}

export const angle = ([x0, y0]: Vector, [x1, y1]: Vector = [0, 0]): number => {
  return Math.atan2(y1 * x0 - y0 * x1, x0 * x1 + y0 * y1);
};

export function normalize([x0, y0]: Vector): Vector {
  const l = distance([x0, y0]);
  return [x0 / l, y0 / l];
}

export function polarToCartesian(r: number, theta: number): Vector {
  const x = Math.cos(theta) * r;
  const y = Math.sin(theta) * r;
  return [x, y];
}

export function cartesianToPolar([x, y]: Vector): [number, number] {
  const r = distance([x, y]);
  const theta = Math.atan2(y, x);

  return [r, theta];
}

export function parallel(v1: Vector, v2: Vector, precision = 1e-9): boolean {
  const V1xV2 = crossProduct(v1, v2);

  const xLength = squareLength(v1);
  const yLength = squareLength(v2);

  return V1xV2 * V1xV2 < xLength * yLength * precision * precision;
}
