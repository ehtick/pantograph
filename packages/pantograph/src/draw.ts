import { Vector } from "./definitions";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { Line } from "./models/segments/Line";
import { Segment } from "./models/segments/Segment";
import { TransformationMatrix } from "./models/TransformationMatrix";
import {
  polarToCartesian,
  DEG2RAD,
  subtract,
  sameVector,
} from "./vectorOperations";

function closeSegments(segments: Segment[]) {
  if (!segments.length) throw new Error("No segments to close");
  const firstSegment = segments[0];
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const lastSegment = segments.at(-1)!;

  if (sameVector(firstSegment.firstPoint, lastSegment.lastPoint))
    return segments;
  return [
    ...segments,
    new Line(lastSegment.lastPoint, firstSegment.firstPoint),
  ];
}

function loopySegmentsToDiagram(segments: Segment[]) {
  // Here we will need to do our best to fix cases where the drawing is
  // broken in some way (i.e. self-intersecting loops)

  return new Diagram([new Figure(new Loop(segments))]);
}

export class DrawingPen {
  protected pointer: Vector;
  protected firstPoint: Vector;
  protected pendingSegments: Segment[];

  constructor(origin: Vector = [0, 0]) {
    this.pointer = origin;
    this.firstPoint = origin;

    this.pendingSegments = [];
  }

  movePointerTo(point: Vector): this {
    if (this.pendingSegments.length)
      throw new Error(
        "You can only move the pointer if there is no segment defined"
      );

    this.pointer = point;
    this.firstPoint = point;
    return this;
  }

  protected saveSegment(segment: Segment) {
    this.pendingSegments.push(segment);
    return this;
  }

  lineTo(point: Vector): this {
    const segment = new Line(this.pointer, point);
    this.pointer = point;
    return this.saveSegment(segment);
  }

  line(xDist: number, yDist: number): this {
    return this.lineTo([this.pointer[0] + xDist, this.pointer[1] + yDist]);
  }

  vLine(distance: number): this {
    return this.line(0, distance);
  }

  hLine(distance: number): this {
    return this.line(distance, 0);
  }

  vLineTo(yPos: number): this {
    return this.lineTo([this.pointer[0], yPos]);
  }

  hLineTo(xPos: number): this {
    return this.lineTo([xPos, this.pointer[1]]);
  }

  polarLineTo([r, theta]: Vector): this {
    const angleInRads = theta * DEG2RAD;
    const point = polarToCartesian(r, angleInRads);
    return this.lineTo(point);
  }

  polarLine(distance: number, angle: number): this {
    const angleInRads = angle * DEG2RAD;
    const [x, y] = polarToCartesian(distance, angleInRads);
    return this.line(x, y);
  }

  tangentLine(distance: number): this {
    const previousCurve = this.pendingSegments.at(-1);

    if (!previousCurve)
      throw new Error("You need a previous segment to sketch a tangent line");

    const [xDir, yDir] = previousCurve.tangentAtLastPoint;
    return this.line(xDir * distance, yDir * distance);
  }

  close(): Diagram {
    const segments = closeSegments(this.pendingSegments);
    return loopySegmentsToDiagram(segments);
  }

  closeWithMirror(): Diagram {
    if (!this.pendingSegments.length) throw new Error("No segments to close");

    const firstSegment = this.pendingSegments[0];
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const lastSegment = this.pendingSegments.at(-1)!;

    const mirrorVector = subtract(
      lastSegment.lastPoint,
      firstSegment.firstPoint
    );
    const mirrorTranform = new TransformationMatrix().mirrorLine(
      mirrorVector,
      firstSegment.firstPoint
    );

    const mirroredSegments = this.pendingSegments.map((segment) =>
      segment.transform(mirrorTranform).reverse()
    );
    mirroredSegments.reverse();

    return loopySegmentsToDiagram([
      ...this.pendingSegments,
      ...mirroredSegments,
    ]);
  }
}

export function draw(origin: Vector = [0, 0]): DrawingPen {
  return new DrawingPen(origin);
}
