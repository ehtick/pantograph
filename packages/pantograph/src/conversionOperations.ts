import { Segment } from "./models/segments/Segment";
import { Arc } from "./models/segments/Arc";
import { Line } from "./models/segments/Line";
import { QuadraticBezier } from "./models/segments/QuadraticBezier";
import { CubicBezier } from "./models/segments/CubicBezier";
import { EllipseArc } from "./models/segments/EllipseArc";
import { approximateBezierCurveAsArcs } from "./algorithms/conversions/bezierToArcs";
import { approximateEllipticalArcAsCubicBeziers } from "./algorithms/conversions/ellipseToBezier";
import { genericConversion } from "./algorithms/conversions/helpers";
import { Diagram, Figure, Loop, Strand } from "./main";

function mapAsArcsAndLines(s: Segment, tolerance = 1e-3): Segment[] {
  if (s instanceof Arc || s instanceof Line) {
    return [s];
  }
  if (s instanceof QuadraticBezier || s instanceof CubicBezier) {
    return approximateBezierCurveAsArcs(s, tolerance);
  }
  if (s instanceof EllipseArc) {
    return approximateEllipticalArcAsCubicBeziers(s).flatMap((s) =>
      approximateBezierCurveAsArcs(s, tolerance),
    );
  }

  throw new Error(`Unknown segment type: ${s}`);
}

function convertToArcsAndLines(shape: Diagram): Diagram;
function convertToArcsAndLines(shape: Figure): Figure;
function convertToArcsAndLines(shape: Loop): Loop;
function convertToArcsAndLines(shape: Strand): Strand;
function convertToArcsAndLines(shape: Segment): Segment[];
function convertToArcsAndLines(
  shape: Diagram | Figure | Loop | Strand | Segment,
  tolerance = 1e-3,
) {
  const mapper = (segment: Segment) => mapAsArcsAndLines(segment, tolerance);
  return genericConversion(shape, mapper);
}

export { convertToArcsAndLines };
