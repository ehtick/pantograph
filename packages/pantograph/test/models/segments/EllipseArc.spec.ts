import { describe, it, expect } from "vitest";
import { RAD2DEG } from "../../../src/main";
import {
  EllipseArc,
  svgEllipse,
} from "../../../src/models/segments/EllipseArc";
import { add, scalarMultiply } from "../../../src/vectorOperations";
import { debugImg, dpnt } from "../../debug";

describe("EllipseArc", () => {
  it("should handle major and minor axes correctly", () => {
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true);
    const arc2 = new EllipseArc([0, 0], [2, 1], [2, 0], 1, 2, -90, true);

    expect(arc.majorRadius).toBe(2);
    expect(arc.minorRadius).toBe(1);

    expect(arc2.majorRadius).toBe(2);
    expect(arc2.minorRadius).toBe(1);

    expect(arc.tiltAngle).toBeCloseTo(0);
    expect(arc2.tiltAngle).toBeCloseTo(0);

    expect(arc.isSame(arc2)).toBe(true);
  });

  it("should handle translations correctly", () => {
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true);
    const translatedArc = arc.translate(1, 1);

    expect(translatedArc.firstPoint).toEqual([1, 1]);
    expect(translatedArc.lastPoint).toEqual([3, 2]);
    expect(translatedArc.center).toEqual([3, 1]);
    expect(translatedArc.majorRadius).toBe(arc.majorRadius);
    expect(translatedArc.minorRadius).toBe(arc.minorRadius);
    expect(translatedArc.tiltAngle).toBe(arc.tiltAngle);

    expect(translatedArc.deltaAngle).toBeCloseTo(arc.deltaAngle);
    expect(translatedArc.firstAngle).toBeCloseTo(arc.firstAngle);
    expect(translatedArc.lastAngle).toBeCloseTo(arc.lastAngle);
  });

  it("should handle rotations correctly", () => {
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true);
    const rotatedArc = arc.rotate(90);

    expect(rotatedArc.firstPoint).toBeVector([0, 0]);
    expect(rotatedArc.lastPoint).toBeVector([-1, 2]);
    expect(rotatedArc.center).toBeVector([0, 2]);
    expect(rotatedArc.tiltAngle * RAD2DEG).toBe(arc.tiltAngle * RAD2DEG + 90);
    expect(rotatedArc.majorRadius).toBeCloseTo(arc.majorRadius);
    expect(rotatedArc.minorRadius).toBeCloseTo(arc.minorRadius);

    expect(rotatedArc.deltaAngle).toBeCloseTo(arc.deltaAngle);
  });

  it("should handle major and minor axes correctly when tilted", () => {
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true).rotate(
      33,
    );
    const arc2 = new EllipseArc([0, 0], [2, 1], [2, 0], 1, 2, -90, true).rotate(
      33,
    );

    expect(arc.majorRadius).toBe(2);
    expect(arc.minorRadius).toBe(1);

    expect(arc2.majorRadius).toBe(2);
    expect(arc2.minorRadius).toBe(1);

    expect(arc.tiltAngle).toBeCloseTo(arc.tiltAngle);

    expect(arc.deltaAngle).toBeCloseTo(arc2.deltaAngle);
    expect(arc.firstAngle).toBeCloseTo(arc2.firstAngle);
    expect(arc.lastAngle).toBeCloseTo(arc2.lastAngle);

    expect(arc.isSame(arc2)).toBe(true);
  });

  it("should handle tilt angle correctly", () => {
    const arc = new EllipseArc([-2, 0], [2, 0], [0, 0], 2, 1, 0, true).rotate(
      -33,
    );
    //debugImg(arc);
    expect(arc.tiltAngle * RAD2DEG).toBeCloseTo(360 - 33);
  });

  it("should reverse correctly", () => {
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true)
      .rotate(33)
      .translate(2, 3);

    expect(arc.reverse().clockwise).toBe(!arc.clockwise);
    expect(arc.reverse().isSame(arc)).toBe(true);
  });

  it("should have stable param to point methods", () => {
    const arc = new EllipseArc([0, 0], [4, 0], [2, 0], 2, 1, 0, true)
      .translateX(-2)
      .rotate(33);

    const p = arc.paramPoint(0.22);
    expect(arc.pointToParam(p)).toBeCloseTo(0.22);
  });

  it("should compute the bounding box correctly", () => {
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false).rotate(
      22,
    );
    const arc2 = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true).rotate(
      22,
    );
    expect(arc.boundingBox).toMatchSnapshot();
    expect(arc2.boundingBox).toMatchSnapshot();
  });

  it("should compute distances to a point correctly", () => {
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false);

    const distantPoint = (param: number, distance: number) => {
      const p0 = arc.paramPoint(param);
      const p1 = arc.normalAt(p0);
      return add(scalarMultiply(p1, distance), p0);
    };

    expect(arc.distanceFrom([2, 0])).toBeCloseTo(1);
    expect(arc.distanceFrom(distantPoint(0, 0.8))).toBeCloseTo(0.8);
    expect(arc.distanceFrom(distantPoint(0.2, 0.8))).toBeCloseTo(0.8);
    expect(arc.distanceFrom(distantPoint(0.4, 0.8))).toBeCloseTo(0.8);
    expect(arc.distanceFrom(distantPoint(0.6, 0.8))).toBeCloseTo(0.8);
    expect(arc.distanceFrom(distantPoint(0.8, 8))).toBeCloseTo(8);
    expect(arc.distanceFrom(distantPoint(0.9, 0.8))).toBeCloseTo(0.8);

    expect(arc.distanceFrom(distantPoint(0.1, -0.1))).toBeCloseTo(0.1);
    expect(arc.distanceFrom(distantPoint(0.2, -0.1))).toBeCloseTo(0.1);
    expect(arc.distanceFrom(distantPoint(0.4, -0.1))).toBeCloseTo(0.1);
    expect(arc.distanceFrom(distantPoint(0.6, -0.1))).toBeCloseTo(0.1);
    expect(arc.distanceFrom(distantPoint(0.8, -0.1))).toBeCloseTo(0.1);
    expect(arc.distanceFrom(distantPoint(0.9, -0.1))).toBeCloseTo(0.1);
  });

  describe("split arcs", () => {
    it("handles the case for a simple split", () => {
      const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false);
      const p = (theta: number) => arc.paramPoint(theta);

      const checkForValue = (theta: number) => {
        const [arc1, arc2] = arc.splitAt(p(theta));
        expect(arc1?.firstPoint).toBeVector([0, 0]);
        expect(arc1?.lastPoint).toBeVector(p(theta));
        expect(arc2?.firstPoint).toBeVector(p(theta));
        expect(arc2?.lastPoint).toBeVector([2, 1]);
      };

      checkForValue(0.0001);
      checkForValue(0.1);
      checkForValue(0.5);
      checkForValue(0.95);
      checkForValue(0.9999);
    });

    it("handles the case for a multiple splits", () => {
      const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false);
      const p = (theta: number) => arc.paramPoint(theta);

      const points = [0.5, 0.002, 0.25, 0.999];

      const sortedPoints = [...points].sort((a, b) => a - b);
      sortedPoints.unshift(0);
      sortedPoints.push(1);

      const arcs = arc.splitAt(points.map(p));

      expect(arcs.length).toBe(points.length + 1);

      arcs.forEach((arc, i) => {
        expect(arc?.firstPoint).toBeVector(p(sortedPoints[i]));
        expect(arc?.lastPoint).toBeVector(p(sortedPoints[i + 1]));
      });
    });

    it("handles the case for a multiple splits with two points that are the same", () => {
      const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false);
      const p = (theta: number) => arc.paramPoint(theta);

      const points = [0.5, 0.002, 0.25, 0.999, 0.5 + 1e-12];

      const sortedPoints = [...points.slice(0, -1)].sort((a, b) => a - b);
      sortedPoints.unshift(0);
      sortedPoints.push(1);

      const arcs = arc.splitAt(points.map(p));

      expect(arcs.length).toBe(points.length);

      arcs.forEach((arc, i) => {
        expect(arc?.firstPoint).toBeVector(p(sortedPoints[i]));
        expect(arc?.lastPoint).toBeVector(p(sortedPoints[i + 1]));
      });
    });
  });
});

describe("svgEllipse", () => {
  it("draws the correct ellipse with the correct attributes", () => {
    const arc = svgEllipse([0, 2], [2, 1], 3, 1, 0, true, false);

    //debugImg([arc, dpnt([0, 0])]);
    expect(arc).toMatchSnapshot();
  });

  it("changes the tilt angle as expected", () => {
    const arc = svgEllipse([0, 0], [42.42, 42.42], 30, 50, 45, false, true);

    //debugImg([arc]);
    expect(arc).toMatchSnapshot();
  });
});
