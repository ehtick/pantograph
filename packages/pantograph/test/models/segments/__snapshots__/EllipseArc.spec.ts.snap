// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EllipseArc > should compute the bounding box correctly 1`] = `
BoundingBox {
  "xMax": 3.7461957891036173,
  "xMin": -0.000001,
  "yMax": 1.9412671636294256,
  "yMin": -0.4428407899657773,
}
`;

exports[`EllipseArc > should compute the bounding box correctly 2`] = `
BoundingBox {
  "xMax": 1.4797621157176628,
  "xMin": -0.037460370836467444,
  "yMax": 1.6763980413986113,
  "yMin": -0.000001,
}
`;

exports[`svgEllipse > changes the tilt angle as expected 1`] = `
EllipseArc {
  "_coefficients": null,
  "_focals": [
    [
      -6.459860961089884,
      48.87986096108989,
    ],
    [
      50.108681533833916,
      -7.688681533833915,
    ],
  ],
  "_linearExentricity": 40,
  "center": [
    21.824410286372014,
    20.595589713627987,
  ],
  "clockwise": true,
  "firstPoint": [
    0,
    0,
  ],
  "lastPoint": [
    42.42,
    42.42,
  ],
  "majorRadius": 50,
  "minorRadius": 30,
  "precision": 0.000001,
  "segmentType": "ELLIPSE_ARC",
  "tiltAngle": 2.356194490192345,
}
`;

exports[`svgEllipse > draws the correct ellipse with the correct attributes 1`] = `
EllipseArc {
  "_coefficients": null,
  "_focals": [
    [
      1.8332406094626617,
      1.0566252188258825,
    ],
    [
      -3.823613640029719,
      1.0566252188258825,
    ],
  ],
  "_linearExentricity": 2.8284271247461903,
  "center": [
    -0.9951865152835286,
    1.0566252188258825,
  ],
  "clockwise": false,
  "firstPoint": [
    0,
    2,
  ],
  "lastPoint": [
    2,
    1,
  ],
  "majorRadius": 3,
  "minorRadius": 1,
  "precision": 0.000001,
  "segmentType": "ELLIPSE_ARC",
  "tiltAngle": 0,
}
`;
